README.md
# libdio

Input / Output library for the PRG1F course. Smart abstractions to ask for user inputs, and display things to the console (colors, display grid).

## Usage

To use this library in your code, you have to add the following lines in your `CMakeLists.txt`:

```cmake
include(FetchContent) # once in the project to include the module

FetchContent_Declare(libdio
        GIT_REPOSITORY https://github.com/HEIGVD-PRG1-F-2022/libdio.git
        GIT_TAG v0.2.2
        )
FetchContent_MakeAvailable(libdio)

# Here are your other declarations, like `add_executable`

# Replace `PROJECT` with the name of your executable
target_link_libraries(PROJECT PRIVATE libdio)
```

Now you can do the following to include the library in your files:

```c++
#include <libdio/display.h>
#include <libdio/input.h>
```

## Features
See a code example [here](example/main.cpp).

### Display text and colors
The Display class can be used with or without an object.

**Normal text**  
To display a text in white, simply use the method `showText()`:
```cpp
Display::showText("Hello !!");
```

**Support of colors**  
libdio supports ANSI colors, and a few common colors are included. Here is the full list and their ANSI codes
```cpp
enum Colors {
	WHITE = 15,
	YELLOW = 226,
	ORANGE = 214,
	RED = 196,
	AQUA = 122,
	PINK = 207,
	GREEN = 40,
	BLUE = 21,
	BLACK = 16
};
```
And a visual test of these colors:  
![colors.png](imgs/colors.png)

But you can cast any integer value of an ANSI color to get a custom color that is not listed above. For example with the color 148:
```cpp
Display::showText(msg, (Display::Colors) 148);
```
![custom-colors.png](imgs/custom-colors.png)

All the following ANSI codes can be used!
![8bits-colors-wikipedia.png](imgs/8bits-colors-wikipedia.png)
*Source: [8 bit ANSI codes on Wikipedia](https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit)*

You want a quick view of all colors in your terminal ? Use this snippet:  
```cpp
for (int i = 0; i < 256; i++) {
	Display::showText("Text in custom color (code = " + to_string(i) + ")", (Display::Colors) i);
}
```

**Text partially coloured**  
![partial-coloured-text.png](imgs/partial-coloured-text.png)
Concatenate a normal text with an escaped string generated by `setTextColor`.
```cpp
Display::showText("Hello " + Display::setTextColor("world !", Display::Colors::GREEN));
```

**Changing cursor position**  
To move the cursor to a given position you can currently save the position and jump to it later:

This is useful to print and update boards smoothly. Just save the cursor position before printing the board, and go back to it before printing it again.

You can do other cool things with it, let's rewrite the last line for example after 1 second:
```cpp
Display::saveCursorPosition();
Display::showText("Step 1", Display::Colors::GREEN);
std::this_thread::sleep_for(1000ms);
Display::goBackToCursorPosition();
Display::showText("Step 2", Display::Colors::RED);
std::this_thread::sleep_for(1000ms);
```

You can create a nice progress counter with percentage and time...
```cpp
int counter = 0;
int progress = 1;
do {
	Display::saveCursorPosition();
	Display::showText("Waiting " + to_string(counter) + "ms. " + to_string(progress) + "%.");
	std::this_thread::sleep_for(100ms);
	counter += 100;           //increase the sleep time
	progress += progress % 10;//increase progress counter for example
	
	//Go back at the start of the line only if this is not the last iteration
	if (progress < 100) {
		Display::goBackToCursorPosition();
	}
} while (progress <= 100);
```

**Advanced output**  
By creating a Display object, you have access to more special methods:
```cpp
Display d;
```

### Display grids
You can display a simple grid of strings:
```cpp
vector<vector<string>> numbers = {
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
};

Display d();
d.DisplayGrid(numbers, true, true);
```

3 parameters
- grid: vector of vector of string
- border: enable the double border
- separation: enable separation inside cell (not working)

![simple-bordered-grid.png](imgs/simple-bordered-grid.png)

And without any border:
```cpp
vector<vector<string>> numbers = {
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
		{"1", "2", "1", "5", "2", "6", "2", "1"},
};

Display d;
    d.DisplayGrid(numbers, false, false);
```

![grid-without-border.png](imgs/grid-without-border.png)

## API reference
Prototypes of all methods included in this library.

Display class:
```cpp
enum Colors {
	WHITE = 15,
	YELLOW = 226,
	ORANGE = 214,
	RED = 196,
	AQUA = 122,
	PINK = 207,
	GREEN = 40,
	BLUE = 21,
	BLACK = 16
};

static std::string setTextColor(std::string text, Colors color);

static void saveCursorPosition();

static void goBackToCursorPosition();

static void showText(std::string text, Colors color = Colors::WHITE);

explicit Display(Colors defaultTextColor = Colors::WHITE);

void setTextColor(Colors color);

void show();

void show(std::string text);

void setContent(std::string string);

void clear();

void DisplayGrid(const std::vector<std::vector<std::string>> &grid,
					bool border = true, bool insideSeparation = true);
```

Input class:
```cpp
template<typename T>
T userInput(const std::string &msg,
            const std::string &err = "Wrong type, please retry");

/**
 * User input with range clamp. This function will not return a value lower than
 * min and higher than max.
 * @param msg message to display to the user
 * @param min minimum value, inclusive
 * @param max maximum value, inclusive
 * @param err error message in case of wrong type
 */
template<typename T>
T userInputRange(const std::string &msg, T min, T max, const std::string &bad_range,
                 const std::string &err = "Wrong type, please retry");
```

## Develop libdio locally

If you want to change how `libdio` works in your local directory, you can do the following:

1. Clone the libdio repo locally
2. Change the code for `CMakeLists.txt` to the following:

```cmake
add_subdirectory(libdio)
# Here are your other declarations, like `add_executable`

# Replace `PROJECT` with the name of your executable
target_link_libraries(PROJECT PRIVATE libdio)
```

Please don't forget to create a pull-request with your changes!

## Images

(Les images ont été prises durant la beta et ne sont pas représentatives du produit final)

![image](https://user-images.githubusercontent.com/46396184/202451747-80dc852e-b7df-4511-9a0b-6907789343b7.png)

![image](https://user-images.githubusercontent.com/46396184/202451486-d65ad651-b26e-4ddf-8b79-2c10e40fdb47.png)


CHANGELOG.md
2022-11-30 - v0.2
    * Rewrite of display

2022-11-23 - v0.1
    * Initial commit
    * Made by Florian Duruz and Aellen Quentin for the course PRG1-f at HEIG/VD - https://heig-vd.ch/


src/display.cpp
// Original authors:
// - Florian Duruz
// - Aellen Quentin

#include <cstdlib>
#include <iostream>
#include <regex>
#include <utility>

#ifdef _WIN32

#include <windows.h>

int enableVirtualTerminalProcessing() {

    HANDLE stdOut = GetStdHandle(STD_OUTPUT_HANDLE);

    if (stdOut != INVALID_HANDLE_VALUE) {
        DWORD mode = 0;
        if (GetConsoleMode(stdOut, &mode)) {
            mode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
            if (SetConsoleMode(stdOut, mode)) {
                return 0;
            }
        }
    }

    return GetLastError();
}

#endif

#include "libdio/display.h"

using namespace std;

const string TOP_LEFT_PIPE = "╔";
const string TOP_RIGHT_PIPE = "╗";
const string HORIZONTAL_PIPE = "═";
const string HORIZONTAL_UP_PIPE = "╩";
const string HORIZONTAL_DOWN_PIPE = "╦";
const string VERTICAL_PIPE = "║";
const string VERTICAL_LEFT_PIPE = "╣";
const string VERTICAL_RIGHT_PIPE = "╠";
const string BOT_LEFT_PIPE = "╚";
const string BOT_RIGHT_PIPE = "╝";
const string CROSS_PIPE = "╬";

namespace Display {

    // Removes all ANSI sequences for counting characters.
    string nonANSI(const string &s);

    // Splits a string along the 'sep' string in as many pieces as possible.
    vector<string> split(const string &s, const string &sep);

    DString::DString(const std::string &s) {
        this->assign(s);
    }

    DString::DString(Display::Color c) {
        setColor(c);
    }

    DString &DString::setColor(Color color) {
        *this += "\x1b[38;5;" + to_string(int(color)) + "m";
        return *this;
    }

    DString &DString::operator<<(const DString &obj) {
        *this += obj;
        return *this;
    }
    DString &DString::operator<<(const std::string &obj) {
        *this += obj;
        return *this;
    }
    DString &DString::operator<<(const char obj[]) {
        *this += obj;
        return *this;
    }
    DString &DString::operator<<(const Color &col) {
        setColor(col);
        return *this;
    }

    DString &DString::print() {
        cout << *this;
        reset();
        return *this;
    }

    DString &DString::reset(Color color) {
        clear();
        setColor(color);
        return *this;
    }

    DString &DString::resetColor() {
        return *this << "\x1b[0m";
    }

    DString &DString::saveCursorPosition() {
        // While the first is the official ANSI code for saving the cursor position,
        // the latter works in most unix terminals, but not MacOS.
#ifdef __WIN32__
        *this += "\x1b[s";
#elif __linux__
        *this += "\x1b"
                 "7";
#endif
        return *this;
    }

    DString &DString::goBackToCursorPosition() {
// While the first is the official ANSI code for saving the cursor position,
// the latter works in most unix terminals, but not MacOS.
#ifdef __WIN32__
        *this += "\x1b[u";
#elif __linux__
        *this += "\x1b"
                 "8";
#endif
        return *this;
    }

    DString &DString::clearScreen() {
        *this += "\x1b[2J\x1b[H";
        return *this;
    }

    size_t DString::max_width() {
        size_t max_width = 0;
        for (const auto &line: split(*this, "\n")) {
            max_width = max(max_width, nonANSI(line).size());
        }
        return max_width;
    }

    size_t DString::count_lines() {
        return size_t(std::count_if(begin(), end(), [](char c) -> bool { return c == '\n'; }));
    }

    vector<string> split(const string &s, const string &sep) {
        vector<string> ret;
        size_t last = 0;
        size_t next = 0;
        while ((next = s.find(sep, last)) != string::npos) {
            ret.push_back(s.substr(last, next - last));
            last = next + 1;
        }
        ret.push_back(s.substr(last));
        return ret;
    }

    // Returns the string without any ANSI sequences
    string nonANSI(const string &s) {
        return regex_replace(s, regex("\x1b.*?[a-zA-Z]"), "");
    }

    void init() {
#ifdef _WIN32
        enableVirtualTerminalProcessing();
        SetConsoleOutputCP(CP_UTF8);
#endif
    }

    void systemClearScreen() {
#ifdef _WIN32
        system("cls");
#elif __unix__
        system("clear");
#endif
    }

    void saveCursorPosition() {
        DString().saveCursorPosition().print();
    }

    void restoreCursorPosition() {
        DString().goBackToCursorPosition().print();
    }

    void clearScreen() {
        DString().clearScreen().print();
    }

    string drawFrame(string first, const string &middle, const string &last, const string &space, size_t cells, size_t width) {
        string line = move(first);
        string spaces;
        for (size_t i = 0; i <= width + 1; i++) {
            spaces += space;
        }
        for (size_t i = 1; i < cells; i++) {
            line += spaces + middle;
        }
        return line + spaces + last;
    }

    string drawInBetween(size_t cells, size_t width) {
        return drawFrame(VERTICAL_RIGHT_PIPE, CROSS_PIPE, VERTICAL_LEFT_PIPE, HORIZONTAL_PIPE, cells, width);
    }

    string drawTopFrame(size_t cells, size_t width) {
        return drawFrame(TOP_LEFT_PIPE, HORIZONTAL_DOWN_PIPE, TOP_RIGHT_PIPE, HORIZONTAL_PIPE, cells, width);
    }

    string drawBottomFrame(size_t cells, size_t width) {
        return drawFrame(BOT_LEFT_PIPE, HORIZONTAL_UP_PIPE, BOT_RIGHT_PIPE, HORIZONTAL_PIPE, cells, width);
    }

    template<>
    DString displayGrid<DString>(const vector<vector<DString>> &grid, bool show_grid) {
        size_t max_width = 0, max_height = 0;
        for (const auto &line: grid) {
            for (auto cell: line) {
                max_width = max(max_width, cell.max_width());
                max_height = max(max_height, cell.count_lines());
            }
        }
        size_t cells = grid[0].size();

        DString out;

        if (show_grid) {
            out << drawTopFrame(cells, max_width) + "\n";
        }

        for (size_t y = 0; y < grid.size(); ++y) {
            if (show_grid) {
                if (y != 0) {
                    out << drawInBetween(cells, max_width) + "\n";
                }
                out << Color::WHITE << VERTICAL_PIPE << " ";
            }
            for (size_t x = 0; x < grid.at(0).size(); ++x) {
                DString cell = grid.at(y).at(x);
                while (cell.count_lines() < max_height) {
                    cell << "\n";
                }
                while (cell.max_width() < max_width) {
                    cell << " ";
                }
                out << cell;

                out << " ";
                if (show_grid && x != grid.at(0).size() - 1) {
                    out << Color::WHITE << VERTICAL_PIPE << " ";
                }
            }
            if (show_grid) {
                out << Color::WHITE << VERTICAL_PIPE;
            }
            out << "\n";
        }
        if (show_grid) {
            out << drawBottomFrame(cells, max_width) + "\n";
        }
        return out;
    }

    template<>
    DString displayGrid<string>(const vector<vector<string>> &grid, bool show_grid) {
        return displayGridConvert<string>(
                grid, [](const string &val) -> DString { return DString() << val; }, show_grid);
    }
}// namespace Display


example/main.cpp
#include <functional>
#include <libdio/display.h>
#include <libdio/input.h>
#include <thread>
#include <vector>

using namespace std;

int main() {
    Display::init();

    Display::DString display(Display::Color::PINK);
    Display::systemClearScreen();
    display.clearScreen();
    display << "Test\n";
    display.print();

    display.setColor(Display::Color::GREEN) << "In green\n";
    display.print();

    Display::saveCursorPosition();
    display << "Hello World!\nNext line\n";
    display.print();

    this_thread::sleep_for(1000ms);

    Display::restoreCursorPosition();
    display.setColor(Display::Color::RED);
    display << "Goodbye world!\nThis is red\n";
    display.print();

    display << "HAL 2000\n";
    cout << display;

    vector<vector<string>> grid(3, vector<string>(3, "0"));
    cout << Display::displayGrid(grid, false);

    grid.at(1).at(1) = "11";
    Display::displayGridConvert<string>(grid, [](const string &s) -> Display::DString {
        return Display::DString(s == "11" ? Display::Color::GREEN : Display::Color::RED) << s;
    }).print();

    int age = userInputRange<int>("How old are you?", 0, 200, "Bad entry");

    cout << (Display::DString() << "You are " << Display::Color::RED << age << Display::Color::WHITE << " years old\n");

    return EXIT_SUCCESS;
}

include/libdio/input.h
#ifndef LIBDIO_INPUT_H
#define LIBDIO_INPUT_H

#include <string>

/**
 * User input
 * @param msg message to display to the user
 * @param err error message in case of wrong type
 */
template<typename T>
T userInput(const std::string &msg,
            const std::string &err = "Wrong type, please retry");

/**
 * User input with range clamp. This function will not return a value lower than
 * min and higher than max.
 * @param msg message to display to the user
 * @param min minimum value, inclusive
 * @param max maximum value, inclusive
 * @param err error message in case of wrong type
 */
template<typename T>
T userInputRange(const std::string &msg, T min, T max, const std::string &bad_range,
                 const std::string &err = "Wrong type, please retry");

#include "../../src/input.tpp"

#endif


include/libdio/display.h
// Original authors:
// - Florian Duruz
// - Aellen Quentin

#ifndef LIBDIO_DISPLAY_H
#define LIBDIO_DISPLAY_H

#include <functional>
#include <iostream>
#include <string>
#include <vector>
/**
 * Display library to be used in the class PRG1-F.
 * This is a work in progress - PRs are very welcome.
 */
namespace Display {

    /**
     * Common colors to be used.
     */
    enum class Color {
        WHITE = 15,
        YELLOW = 226,
        ORANGE = 214,
        RED = 196,
        AQUA = 122,
        PINK = 207,
        GREEN = 40,
        BLUE = 21,
        BLACK = 16
    };

    // DString represents a colored string. Upon creation, it is initialised with
    // a white text color.
    // It inherits the string class, so all the methods for string are also available
    // to `DString`.
    // In addition, you can use the `<<` operator to add new strings or a new color
    // to a `DString`.
    // Once you're done, you can either call `print()` or output the `DString` to `cout`.
    // This allows for a much faster output, specifically if you create a string out
    // of many elements.
    class DString : public std::string {
    public:
        explicit DString(const std::string &s);

        /**
         * Instantiate a new DString.
         * @param c the color of the string.
         */
        explicit DString(Color c = Color::WHITE);

        /**
         * Sets the color for all following text.
         * @param color for all upcoming text.
         * @return a reference to DString for chaining
         */
        DString &setColor(Color color);

        /**
         * Use DString as a stream
         * @return a reference to DString
         */
        DString &operator<<(const DString &obj);

        /**
         * Use DString as a stream
         * @return a reference to DString
         */
        DString &operator<<(const std::string &obj);

        /**
         * Use DString as a stream
         * @return a reference to DString
         */
        DString &operator<<(const char obj[]);

        /**
         * Use DString as a stream
         * @return a reference to DString
         */
        DString &operator<<(const Color &col);

        /**
         * Use DString as a stream
         * @return a reference to DString
         */
        template<typename T>
        DString &operator<<(T obj);

        /**
         * Prints the string and clears DString.
         * @return a reference to the DString for chaining.
         */
        DString &print();

        /**
         * Resets the string and initialises it with
         * a new color.
         * @return a reference to the DString for chaining.
         */
        DString &reset(Color color = Color::WHITE);

        /// Resets the text color.
        DString &resetColor();

        /**
         * Saves the current cursor position.
         * @return a reference to the DString for chaining.
         */
        DString &saveCursorPosition();

        /**
         * Restores the previously saved cursor position.
         * @return a reference to the DString for chaining.
         */
        DString &goBackToCursorPosition();

        /**
         * Adds a clear string command.
         * @return a reference to the DString for chaining.
         */
        DString &clearScreen();

        /**
         * Returns the maximum line-length of this string.
         * @return the longest line of this string without the ANSI codes
         */
        size_t max_width();

        /**
         * Returns the number of lines in this string
         * @return number of lines
         */
        size_t count_lines();
    };

    /**
     * Initialises the screen - mostly needed for Windows.
     * But for cross-compatibility it should be called anyway.
     */
    void init();

    /**
     * Saves the current cursor position.
     */
    void saveCursorPosition();

    /**
     * Restores a previously saved cursor position.
     */
    void restoreCursorPosition();

    /**
     * Clears the screen using ANSI codes.
     */
    void clearScreen();

    /**
     * Clears the screen using the `system` command.
     */
    void systemClearScreen();

    /**
     * Displays a grid of a 2D vector. Any type that can be sent to `to_string`, plus
     * `string` and `DString` can be passed.
     * @tparam T compatible types to `to_string`, `string`, and `DString`
     * @param grid the 2D grid
     * @param show_grid whether to show the lines between the cells
     * @return the DString for the grid. It must be `print`ed to be shown.
     */
    template<typename T>
    DString displayGrid(const std::vector<std::vector<T>> &grid, bool show_grid = true);
    template<>
    DString displayGrid<std::string>(const std::vector<std::vector<std::string>> &grid, bool show_grid);
    template<>
    DString displayGrid<DString>(const std::vector<std::vector<DString>> &grid, bool show_grid);

    /**
     * Like `displayGrid` but also takes a closure to convert the type to a DString.
     * @tparam T compatible types to `to_string`, `string`, and `DString`
     * @param grid the 2D grid
     * @param convert a closure that takes T and returns a `DString`
     * @param show_grid whether to show the lines between the cells
     * @return the DString for the grid. It must be `print`ed to be shown.
     */
    template<typename T>
    DString displayGridConvert(const std::vector<std::vector<T>> &grid, std::function<DString(T)> convert, bool show_grid = true);
}// namespace Display

#include "../../src/display.tpp"

#endif// LIBDIO_DISPLAY_H
